src/Data/Text/Internal/Reverse.hs

```diff
@@ -0,0 +1,71 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE UnliftedFFITypes #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+
+{-# OPTIONS_HADDOCK not-home #-}
+
+-- | Implements 'reverse', using efficient C routines by default.
+module Data.Text.Internal.Reverse (reverse) where
+
+#if !defined(PURE_HASKELL)
+import GHC.Exts as Exts
+import Control.Monad.ST.Unsafe (unsafeIOToST)
+import Foreign.C.Types (CSize(..))
+#else
+import Control.Monad.ST (ST)
+import Data.Text.Internal.Encoding.Utf8 (utf8LengthByLeader)
+#endif
+import Prelude hiding (reverse)
+import Data.Text.Internal (Text(..))
+import Control.Monad.ST (runST)
+import qualified Data.Text.Array as A
+
+-- | /O(n)/ Reverse the characters of a string.
+--
+-- Example:
+--
+-- $setup
+-- >>> T.reverse "desrever"
+-- "reversed"
+reverse ::
+#if defined(ASSERTS)
+  HasCallStack =>
+#endif
+  Text -> Text
+#if defined(PURE_HASKELL)
+reverse (Text ba off len) = runST $ do
```

Comment:
```suggestion
reverse v@(Text _ _ 0) = v
reverse (Text ba off len) = runST $ do
```

Reply:
Ah, missed this suggestion.

This should be done in the 'fast' version as well, shouldn't it?!

Reply:
Yes, let's do it in both versions.

Reply:
I'm concerned that this would be a change to how long an unused ByteArray gets retained. I don't know if there's some general strategy about this or not. I think I should leave it out of this change since it's orthogonal to adding a pure-Haskell implementation.